loader/src/sources/albertsons/scrape-albertsons-stores.js

```diff
@@ -0,0 +1,305 @@
+  if (elapsed < navigationInterval) {
+    await timers.setTimeout(navigationInterval - elapsed);
+  }
+  await page.goto(url, options);
+  lastNavigation = Date.now();
+}
+
+let currentBrowser = null;
+let currentPage = null;
+async function withPage(options, callback) {
+  if (!callback && typeof options === "function") {
+    callback = options;
+    options = undefined;
+  }
+
+  let triesLeft = (options?.retries ?? 0) + 1;
+  while (triesLeft > 0) {
+    triesLeft--;
+
+    if (!currentBrowser) {
+      currentBrowser = await puppeteer.launch({
+        headless: !process.env.DEBUG,
+        args: ["--no-sandbox"],
+        ...options?.browser,
+      });
+    }
+    if (!currentPage) {
+      currentPage = await currentBrowser.newPage();
+    }
+
+    try {
+      return await callback(currentPage);
+    } catch (error) {
+      let retry = false;
+      if (error?.message?.toLowerCase()?.includes("target closed")) {
+        console.error("Browser crashed");
+        currentBrowser = null;
+        currentPage = null;
+        retry = triesLeft > 0;
+      }
+      if (!retry) {
+        throw error;
+      }
+    }
+  }
+}
+
+async function closeBrowser() {
+  if (currentBrowser) await currentBrowser.close();
+
+  currentBrowser = null;
+  currentPage = null;
+}
+
+/**
+ * Select a single element from the page, and assert that the given selector
+ * ONLY matches one element.
+ * @param {puppeteer.Page} page
+ * @param {string} selector
+ * @returns {puppeteer.ElementHandle}
+ */
+async function selectOnly(page, selector) {
+  const elements = await page.$$(selector);
+  assert.equal(elements.length, 1, `There should only be one "${selector}"`);
+  return elements[0];
+}
+
+/**
+ * Crawl the website for a single Albertsons sub-brand, yielding the data
+ * object for each pharmacy location found.
+ * @param {puppeteer.Browser} browser
+ * @param {string} brandHost
+ * @param {number[]} [startIndexes]
+ * @yields {any}
+ */
+async function* crawlBrand(brandHost, startIndexes = []) {
+  console.error(`Crawling ${brandHost}...`);
+
+  const statesListUrl = `https://local.pharmacy.${brandHost}`;
+  yield* crawlListingPage(statesListUrl, 0, startIndexes);
+}
+
+/**
+ * Crawl all the links on a listing page and its sub-pages, yielding the data
+ * object for each pharmacy location found.
+ * @param {puppeteer.Page} page
+ * @param {string} pageUrl
+ * @param {number} [depth] Indicates the current recursion depth.
+ * @param {number[]} [startIndexes] Start with the link at a given index on this
+ *        page and its sub-pages.
+ * @yields {any}
+ */
+async function* crawlListingPage(pageUrl, depth = 0, startIndexes = []) {
+  if (depth > 4) throw new Error(`Too many listing pages deep: "${pageUrl}"`);
+
+  let urls = await withPage({ retries: 1 }, async (page) => {
+    await navigate(page, pageUrl);
+
+    // If the state only has one store, this will actually be a store page.
+    const locationData = await getLocationDataFromStorePage(page);
```

Comment:
Incomplete refactor here?

Reply:
Heh, yep, I messed up that commit. Will fix.

Reply:
Should be fixed in 788916eebe434295f0941c6e87f873c2170b9b05